src/box/key_def.c

```diff
@@ -1219,3 +1219,31 @@ key_validate_parts(const struct key_def *key_def, const char *key,
 	*key_end = key;
 	return 0;
 }
+
+char *
+mp_key_append_part_count(const char *key, uint32_t part_count, char **end)
```

Comment:
This function looks like an internal helper. No need to make it public. I'd even inline it in `error_set_key_with_part_count`.

Moreover, I think that this payload field should be set by `key_validate_parts`. If we did that, we could drop `error_set_key_with_part_count` altogether (inline it in `key_validate_parts`).

Reply:
Dropped `mp_key_append_part_count`.

<details>
<summary>diff</summary>

```diff
diff --git a/src/box/key_def.c b/src/box/key_def.c
index 29681ab05..e1648d55c 100644
--- a/src/box/key_def.c
+++ b/src/box/key_def.c
@@ -1220,30 +1220,18 @@ key_validate_parts(const struct key_def *key_def, const char *key,
 	return 0;
 }
 
-char *
-mp_key_append_part_count(const char *key, uint32_t part_count, char **end)
+void
+error_set_key_with_part_count(const char *key, uint32_t part_count)
 {
 	const char *key_end = key;
 	for (uint32_t i = 0; i < part_count; i++)
 		mp_next(&key_end);
-	uint32_t prefix_size = mp_sizeof_array(part_count);
-	char *key_buf = (char *)xregion_alloc(&fiber()->gc,
-					      key_end - key + prefix_size);
-	char *key_pos = key_buf;
-	key_pos = (char *)mp_encode_array(key_pos, part_count);
-	memcpy(key_pos, key, key_end - key);
-	*end = key_buf + (key_end - key + prefix_size);
-	return key_buf;
-}
-
-void
-error_set_key_with_part_count(const char *key, uint32_t part_count)
-{
 	size_t region_svp = region_used(&fiber()->gc);
-	char *key_array, *key_array_end;
-	key_array = mp_key_append_part_count(key, part_count,
-					     &key_array_end);
+	size_t key_buf_size = key_end - key + mp_sizeof_array(part_count);
+	char *key_buf = (char *)xregion_alloc(&fiber()->gc, key_buf_size);
+	char * key_pos = (char *)mp_encode_array(key_buf, part_count);
+	memcpy(key_pos, key, key_end - key);
 	struct error *e = diag_last_error(diag_get());
-	error_set_mp(e, "key", key_array, key_array_end - key_array);
+	error_set_mp(e, "key", key_buf, key_buf_size);
 	region_truncate(&fiber()->gc, region_svp);
 }
diff --git a/src/box/key_def.h b/src/box/key_def.h
index 03b38fb3b..c41f2303a 100644
--- a/src/box/key_def.h
+++ b/src/box/key_def.h
@@ -1180,14 +1180,6 @@ key_hint(const char *key, uint32_t part_count, struct key_def *key_def)
 	return key_def->key_hint(key, part_count, key_def);
 }
 
-/**
- * Prepend MsgPack array of part_count size to key. Resulting MsgPack
- * is allocated on fiber region. Return the pointer to the beginning
- * of the result. Pointer to the end is returned in key_end.
- */
-char *
-mp_key_append_part_count(const char *key, uint32_t part_count, char **end);
-
 /**
  * Add key as the last error payload field with name "key". The key
  * is MsgPack array without array prefix. The size of the array

```

</details>

I have to keep `error_set_key_with_part_count` here as it used from vinyl also.